platform/linux-generic/odp_ipsec_sad.c

```diff
@@ -446,6 +446,39 @@ static uint32_t esp_block_len_to_mask(uint32_t block_len)
 	return block_len - 1;
 }
 
+/* AR window management initialization */
+static int ipsec_antireplay_init(ipsec_sa_t *ipsec_sa,
+				 const odp_ipsec_sa_param_t *param)
+{
+	uint16_t num_bkts = 0;
+
+	if (param->inbound.antireplay_ws > IPSEC_AR_WIN_SIZE_MAX) {
+		ODP_ERR("AR_WS[%u] not supported\n", param->inbound.antireplay_ws);
+		return -1;
+	}
+
+	ipsec_sa->antireplay = (param->inbound.antireplay_ws != 0);
+	if (!ipsec_sa->antireplay)
+		return 0;
+
+	ipsec_sa->hot.in.win_size = param->inbound.antireplay_ws;
+	/* Window size should be at least IPSEC_AR_WIN_SIZE_MIN */
+	if (ipsec_sa->hot.in.win_size < IPSEC_AR_WIN_SIZE_MIN)
+		ipsec_sa->hot.in.win_size = IPSEC_AR_WIN_SIZE_MIN;
+
+	num_bkts = ROUNDUP_POWER2_U32(ipsec_sa->hot.in.win_size) / IPSEC_AR_WIN_BUCKET_SIZE;
```

Comment:
Since num_bucket gets rounded up like that, the actual anti-replay window could perhaps also be rounded up to the maximum size allowed by the rounded num_bucket value. I do not see a whole lot of benefit in restricting the window to be smaller than what is anyway naturally handled by the checking code without extra effort.

Reply:
> Since num_bucket gets rounded up like that, the actual anti-replay window could perhaps also be rounded up to the maximum size allowed by the rounded num_bucket value. I do not see a whole lot of benefit in restricting the window to be smaller than what is anyway naturally handled by the checking code without extra effort.

The anti-replay window would not be rounded. 
The rounding of the buckets is as you suggested to avoid the integer division.

Reply:
I am not saying that the window is rounded but that the code could be changed so that it gets rounded (not to power of two but the largest window allowed by the number of buckets). In other words, the size of the window (and thus packet reordering tolerance) could be rounded up without it causing additional processing cost.

If the window is not rounded, then (depending on window size) you could have up to half of the buckets unused for keeping track of which packets were received even though the status bits are already there and the code traverses and touches the full bucket array anyway.

Reply:
> I am not saying that the window is rounded but that the code could be changed so that it gets rounded (not to power of two but the largest window allowed by the number of buckets). In other words, the size of the window (and thus packet reordering tolerance) could be rounded up without it causing additional processing cost.

Yes, understood, can be done. This will also avoid the corner case problem of top new bucket clearing the bottom bucket. 

I am trying to give a example to align my understanding with you.

Lets say, user passed window size of 130, and the buckets rounded value is 4 in this case. So we can safely increase the window size internally to 192(64 * num_buckets - 64 = 64*4 -64 =  192).

Hope we are aligned now.

Reply:
> This will also avoid the corner case problem of top new bucket clearing the bottom bucket.

No, that is a separate issue. It is a bug (and not a corner case problem but a fatal bug) in calculation of the needed number of buckets. One the number of minimum number of buckets for a requested window size is known (and hopefully calculated correctly), we can increase the actual window size to what can be handled using that number of buckets.

> Lets say, user passed window size of 130, and the buckets rounded value is 4 in this case. So we can safely increase the window size internally to 192(64 * num_buckets - 64 = 64*4 -64 = 192).

Yes.

Reply:
> > This will also avoid the corner case problem of top new bucket clearing the bottom bucket.
> 
> No, that is a separate issue. It is a bug (and not a corner case problem but a fatal bug) in calculation of the needed number of buckets. One the number of minimum number of buckets for a requested window size is known (and hopefully calculated correctly), we can increase the actual window size to what can be handled using that number of buckets.

Okay, understood. So this means that in the case of the application passed window sizes from 4033 to 4096, the supported window size internally is 4032(4096 - 64), and the number of rounded buckets are 64( the maximum supported by the implementation).

Reply:
An ODP implementation is allowed to increase the actual window size from what was requested by the application but it is not allowed to decrease it. So, if the implementation advertises max window of 4096 packets in IPsec capabilities, it is not allowed to round it down to something smaller. Thus 64 buckets of 64 packets each would not be enough.

Reply:
So the minimum number of buckets required in this case are 64+1=65. After rounding(bucket rounding is to avoid the integer division during the window update processing) it would be increased to 128. So is it okay to increase the max supported buckets to 128.

Reply:
There is a tradeoff between the potential speedup of getting rid of the division and the drawbacks (reserving more memory & touching more memory) of rounding the number of buckets up. With relatively small window sized the rounding probably makes sense, but with large window sizes it may be counter productive. It is hard to tell which approach is better without suitable measurements.

Maybe a better approach would be to try to make the code so that rounding is not needed but integer division is still avoided. For example, the bucket for a packet that falls within the current window would not be calculated as seqno % nb_bucket but as (seqno - bottom_seqno) >> BUCKET_BITS, where bottom_seqno is updated when the buckets are moved. And similarly for the other cases.

Reply:
We will consider your suggestion and check this solution for the integer division optimization. So there are now two optimizations that are to be checked.

- Integer division optimization( with out bucket rounding).
- top sequence number update once per bucket( not per every packet).

As these optimizations will take some more time, would you please allow me to take these two optimizations as part of a separate PR, so that I will update this PR with all your other comments/suggestions. Please suggest.

Reply:
Feel free to leave the optimizations out of the next version of this PR. Let's check after that if we want to include some of them.

Reply:
Thanks Janne.